### ISSUE TITLE
MSVC, Windows Phone and " warning C4805: '|=': unsafe mix of type 'std::size_t' and type 'bool' in operation"

### ISSUE THREAD

**Issue by noloader** [randombit/botan#795]
Sorry about this extra noise. This should be pretty easy to clear with `!!` to convert from an integral to a bool. All compilers I am aware of accept it.

```
cl  /MD /bigobj /FI sdkddkver.h /FI winapifamily.h /DWINAPI_FAMILY=WINAPI_FAMILY_PHONE_APP /DBOTAN_DLL=__declspec(dllexport) /EHs /GR /O2 /W4 /wd4250 /wd4251 /wd4275 /Ibuild\include /nologo /c C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp /Fobuild\obj\lib\modes_mode_pad.obj
mode_pad.cpp
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(60): warning C4805: '|=': unsafe mix of type 'std::size_t' and type 'bool' in operation
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(66): warning C4805: '&': unsafe mix of type 'int' and type 'bool' in operation
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(101): warning C4805: '|=': unsafe mix of type 'std::size_t' and type 'bool' in operation
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(107): warning C4805: '&': unsafe mix of type 'int' and type 'bool' in operation
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(180): warning C4805: '|=': unsafe mix of type 'std::size_t' and type 'bool' in operation
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(186): warning C4805: '&': unsafe mix of type 'std::size_t' and type 'bool' in operation
build\include\botan/internal/ct_utils.h(91): warning C4805: '|=': unsafe mix of type 'bool' and type 'int' in operation
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(60): note: see reference to function template instantiation 'T Botan::CT::expand_mask<bool>(T)' being compiled
        with
        [
            T=bool
        ]
build\include\botan/internal/ct_utils.h(91): warning C4805: '|': unsafe mix of type 'bool' and type 'int' in operation
build\include\botan/internal/ct_utils.h(91): warning C4800: 'int': forcing value to bool 'true' or 'false' (performance warning)
build\include\botan/internal/ct_utils.h(93): warning C4800: 'int': forcing value to bool 'true' or 'false' (performance warning)
build\include\botan/internal/ct_utils.h(118): warning C4244: 'return': conversion from 'int' to 'uint8_t', possible loss of data
C:\botan-src\src\lib\modes\mode_pad\mode_pad.cpp(66): note: see reference to function template instantiation 'T Botan::CT::is_equal<uint8_t>(T,T)' being compiled
        with
        [
            T=uint8_t
        ]
```

**Comment by randombit** [comment]
Thanks for the report.

> This should be pretty easy to clear with !! to convert from an integral to a bool.

Actually that would be very bad in this context since it's known that `!some_predicate` syntax is compiled by some compilers (including GCC) to use conditional jumps, which we are attempting to avoid here. This warning indicates to me VC may be compiling this code in a way that has jumps depending on the loop predicate. We would have no way to detect this automatically, since we rely on valgrind for testing the const time assertions.

These warnings all seem to be due to `expand_mask(x < y)`, and already the const-time utils have predicates for less than, I will convert the mode unpadding code to use this function. In a further twist the util function itself just does `expand_mask(x < y)` (that's the warning from ct_utils.h) instead of something more complex that the compiler cannot so easily optimize out; that was sufficient with GCC but maybe not for VC. I'll address this also.

**Comment by randombit** [comment]
I made a brief attempt at this but after running into problems I need to put more work in the const time utils than I have time for before 2.0.0. If you are curious, you could check if VC generated asm that included conditional jumps in these code blocks.

Fortunately the utils for side channels is a library-internal header, so there will be no user-visible change when I rewrite ct_utils.h (2.0.1 hopefully?). This issue will be addressed at that time.

**Comment by noloader** [comment]
> Actually that would be very bad in this context since it's known that !some_predicate syntax is compiled by some compilers (including GCC) to use conditional jumps,

Oh my, sorry about that. I did not realize constant-timeness was a goal.

Since Botan requires C++11, and that requires Visual Studio 2010 on Windows, you could perform the following. It side steps a rewrite and will clear a lot of the noise; and the push and pop avoids cross-pollinating into user's headers.

Testing the patch below shows it clears _a lot_ of warnings generated from that header. I'm guessing its in the 50's to 100 or so.

```
// Top of the header file
#ifdef _MSC_VER
# pragma warning(push)
# pragma warning(disable: 4805)
#endif

namespace Botan {
...
} // namespace

// Bottom of the header file
#ifdef _MSC_VER
# pragma warning(pop)
#endif
```

**Comment by noloader** [comment]
@randombit, did you want this closed? I don't feel comfortable asking for a code change due to the constant-time requirements. I don't think its fair to pollute the bug tracker with it.

**Comment by randombit** [comment]
No leave it open, one project goal is to be as warning-clean as possible, and I still think the warning may be pointing to a problem with how VC is compiling our const time utils. I hope to address these warnings in 2.1.

**Comment by randombit** [comment]
I think this was fixed in df4287c3c76
